### ISSUE TITLE
Elasticsearch sort not work in "es.query"

### ISSUE THREAD

**Issue by arman1371** [elastic/elasticsearch-hadoop#689]
when set `es.query` to:

``` json
{
  "query" : {
    "bool" : {
      "filter" : {
        "range" : {
          "collectionTimestamp" : {
            "from" : "2016-01-04T00:00Z",
            "to" : "2016-01-04T00:00:10Z",
            "include_lower" : true,
            "include_upper" : true
          }
        }
      },
      "minimum_should_match" : "1"
    }
  },
  "sort" : [ {
    "vendor" : {
      "order" : "asc",
      "missing" : "_first"
    }
  } ]
}
```

the query partition works good but results are not ordered.

If order by spark, spark first get all elements and then sort them to get just 10 first elements and this is very slow if results are large.
but if use elasticsearch sort, when spark gets 10 first elements, it finishes the work and return the results.

**Comment by bsao** [comment]
I am facing same problem! There is any configuration or is it a bug?

**Comment by costin** [comment]
Sorting requires a central node that retrieves the results, sorted, from all nodes, sorts them again and then returns them. This is not something part of es-hadoop since all its calls are parallelized and further more, Hadoop/spark don't allow such a job monitoring entity.
Further more an RDD is meant to be parallelized, even if it is sorted, operations applied to it will be done out of order so the ending result will ignore the sorting.
In other words, if you _really_ want this (and do note this means Spark as well requires a global lock to sort plus only one partition of the `RDD` to force consumers to use it) you can do so in Spark: require only X docs from each partition, put them in the RDD, sort them and then merge the partition.

**Comment by jpparis-orange** [comment]
Just hit the very same problem today. I'm playing with es.scroll.limit to ramp up the size of tha data spark is ingesting from my ES index. But, I want to sort so that I'm sure the start of my data is always the same.

I do understand Costin's comment on global ordering which needs a unique central node.
Would it be possible that sorting within the ES-hadoop connector just run on a per shard basis? No more central node needed, but a kind of partial order that only depends on shard segmentation.

**Comment by costin** [comment]
Currently ES-Hadoop relies on scan and scroll to retrieve data which means is at efficient as it can be but discards sorting. This typically is not an issue (as mentioned above).
Assuming ES-Hadoop does sorting, this will not be global but rather per _partition_ (that is per shard).
Depending on how the runtime framework works, the data might be shuffled significantly or not; that is the per-shard order might be kept or not, since there are no such guarantees.
In either way, there would be no global ordering and potentially only partially sorted data.
To properly guarantee global sorting there needs to be a "gateway" that sorts everything - this can happen in ES but by doing this, parallelism is killed since the calls would not be per shard, but just to one node that does it.

**Comment by eyaldahari** [comment]
@costin - Is this going to be implemented in the partition (shard) level on future release of the connector like you mentioned in your comment?

**Comment by jbaiera** [comment]
Closing this since the conversation has slowed down and drifted from the original issue. The usage of sorting within an input split is a very niche use case and not one that we have on the immediate roadmap. If there is still reasonable push from the community for it then a new issue should be opened for the feature.

**Comment by ebuildy** [comment]
At least, an error could be thrown to indicate this is not supported.
